WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift

```diff
@@ -24,51 +27,112 @@ final class MediaCollectionCellViewModel {
+    func onDisappear() {
+        guard isVisible else { return }
+        isVisible = false
+        cancelThumbnailRequestIfNeeded()
+    }
+
+    func startPrefetching() {
+        guard !isPrefetchingNeeded else { return }
+        isPrefetchingNeeded = true
+        fetchThumbnailIfNeeded()
+    }
+
+    func cancelPrefetching() {
+        guard isPrefetchingNeeded else { return }
+        isPrefetchingNeeded = false
+        cancelThumbnailRequestIfNeeded()
+    }
+
+    // MARK: - Thumbnail
 
-        requestCount += 1
-        guard requestCount == 1 else {
+    private func fetchThumbnailIfNeeded() {
+        guard isVisible || isPrefetchingNeeded else {
+            return
+        }
+        guard imageTask == nil else {
             return // Already loading
         }
-        let task = Task { [service, media, weak self] in
+        guard getCachedThubmnail() == nil else {
+            return // Already cached  in memory
+        }
+        imageTask = Task { @MainActor [service, media, weak self] in
             do {
-                let image = try await service.image(for: media, preferredSize: targetSize)
-                self?.didFinishLoading(with: image, error: nil)
+                let image = try await service.thumbnail(for: media)
+                self?.didFinishLoading(with: image)
             } catch {
-                self?.didFinishLoading(with: nil, error: error)
+                self?.didFinishLoading(with: nil)
             }
         }
-        imageTask = task
     }
 
-    private func didFinishLoading(with image: UIImage?, error: Error?) {
+    private func cancelThumbnailRequestIfNeeded() {
+        guard !isVisible && !isPrefetchingNeeded else { return }
+        imageTask?.cancel()
+        imageTask = nil
+    }
+
+    private func didFinishLoading(with image: UIImage?) {
         if let image {
             cache.setImage(image, forKey: makeCacheKey(for: media))
         }
-        onLoadingFinished?(image)
-        requestCount = 0
+        if !Task.isCancelled {
+            if let image {
+                onImageLoaded?(image)
+            }
+            imageTask = nil
+        }
     }
 
-    func cancelLoading() {
-        guard requestCount > 0 else { return }
-        requestCount -= 1
-        if requestCount == 0 {
-            imageTask?.cancel()
-        }
+    /// Returns the image from the memory cache.
+    func getCachedThubmnail() -> UIImage? {
+        cache.getImage(forKey: makeCacheKey(for: media))
     }
 
     private func makeCacheKey(for media: Media) -> String {
         "thumbnail-\(media.objectID)"
     }
+
+    // Monitors thumbnails generated by `MediaImportService`.
+    private func didUpdateLocalThumbnail() {
+        guard media.remoteStatus != .sync, media.localThumbnailURL != nil else { return }
+        fetchThumbnailIfNeeded()
+    }
+
+    // MARK: - Status
+
+    private func updateOverlayState() {
+        switch media.remoteStatus {
```

Comment:
It now tracks upload progress directly without even needing `MediaCoordinator`, which allowed me to significantly simplify the code compared to the previous versions.